torchtext/transforms.py

```diff
@@ -309,6 +309,105 @@ def __prepare_scriptable__(self):
+
+    This tokenizer has been trained to treat spaces like parts of the tokens
+    (a bit like sentencepiece) so a word will be encoded differently whether it
+    is at the beginning of the sentence (without space) or not.
+
+    :param encoder_json_path: Path to BPE encoder json file.
+    :type encoder_json_path: str
+    :param vocab_bpe_path: Path to bpe vocab file.
+    :type vocab_bpe_path: str
+    """
+
+    _seperator: torch.jit.Final[str]
+
+    def __init__(
+        self,
+        encoder_json_path: str,
+        vocab_bpe_path: str,
+    ):
+        super().__init__()
+        self._seperator = "\u0001"
+        # load bpe encoder
+        with open(get_asset_local_path(encoder_json_path), "r", encoding="utf-8") as f:
+            bpe_encoder = json.load(f)
+        # load bpe vocab
+        with open(get_asset_local_path(vocab_bpe_path), "r", encoding="utf-8") as f:
+            bpe_vocab = f.read()
+        bpe_merge_ranks = {
+            self._seperator.join(merge_pair.split()): i
+            for i, merge_pair in enumerate(bpe_vocab.split("\n")[1:-1])
+        }
+        # Caching is enabled in Eager mode
+        self.bpe = CLIPEncoderPyBind(bpe_encoder, bpe_merge_ranks,
+                                     self._seperator, bytes_to_unicode(), True)
+
+    @property
+    def is_jitable(self):
+        return isinstance(self.bpe, torch._C.ScriptObject)
+
+    @torch.jit.export
+    def _tokenize(self, text: str) -> List[str]:
+        """Encode text into a list of tokens
+
+        Args:
+            text: An input text string.
+
+        Returns:
+            A list of bpe token ids represents each bpe tokens
+
+        For example: "awesome,awe"
+            --> bpe --> bpe tokens: ["aw", "esome"], [","], ["aw", e]
+            --> bpe encode --> bpe token ids: [707, 5927, 11, 707, 68]
+        """
+        text = text.lower().strip()
+        bpe_token_ids: List[int] = self.bpe.encode(text)
+        bpe_tokens: List[str] = []
+
+        for bpe_token_id in bpe_token_ids:
+            bpe_tokens.append(str(bpe_token_id))
+
+        return bpe_tokens
+
+    def forward(self, input: Any) -> Any:
+        """
+        :param input: Input sentence or list of sentences on which to apply tokenizer.
+        :type input: Union[str, List[str]]
+        :return: tokenized text
+        :rtype: Union[List[str], List[List(str)]]
+        """
+        if torch.jit.isinstance(input, List[str]):
+            tokens: List[List[str]] = []
+            for text in input:
+                tokens.append(self._tokenize(text))
+            return tokens
+        elif torch.jit.isinstance(input, str):
+            return self._tokenize(input)
+        else:
+            raise TypeError("Input type not supported")
+
+    def __prepare_scriptable__(self):
+        r"""Return a JITable tokenizer.
+        """
+        if not self.is_jitable:
+            tokenizer_copy = deepcopy(self)
+            # Disable caching in script mode
```

Comment:
why do we disable caching in script mode? Also do we have some stats on how much we gain performance with caching?

Reply:
Disabling caching in script mode is a safe option. In future, we can design an API that can expose whether model developers want caching in script mode or not.

In script mode the tokenizer is used mostly for production and the cache can technically grow infinitely in such a case. This can lead to OOM errors which will be hard to debug.

Reply:
Hmm, I see. Sounds good. Well, technically we can always limit how much cache we want to keep, but I agree that  that would require some API redesign for user-control etc. We can park this topic until then.